keycloak-operator 26.4.0 missing clusterrole permissions

### Before reporting an issue

- [x] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

_No response_

### Describe the bug

Getting this after updating to 26.4.0

> 2025-09-30 17:50:07,029 WARN [io.jav.ope.pro.eve.EventProcessor] (ReconcilerExecutor-keycloakcontroller-54) Uncaught error during event processing ExecutionScope{ resource id: ResourceID{name='keycloak', namespace='auth'}, version: 22441437} - but ano ther reconciliation will be attempted because a superseding event has been received or another retry attempt is pending.: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1:443/apis/apiextensions.k8s.io /v1/customresourcedefinitions/servicemonitors.monitoring.coreos.com. Message: customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource " customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope. Received status: Status(apiVersion=v1, code=403, details=StatusDetails(causes=[], group=apiextensions.k8s.io, kind=customresourcedefinitions, name=servicemonitors.moni toring.coreos.com, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Forbidden, statu s=Failure, additionalProperties={}).

Editing `keycloak-operator-clusterrole` to add "list" and "watch" permissions to "customresourcedefinitions" fixes it

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"rbac.authorization.k8s.io/v1","kind":"ClusterRole","metadata":{"annotations":{},"labels":{"app.kubernetes.io/name":"keycloak-operator"},"name":"keycloak-operator-clusterrole"},"rules":[{"apiGroups":["apiextensions.k8s.io"],"resources":["customresourcedefinitions"],"verbs":["get"]},{"apiGroups":["config.openshift.io"],"resources":["ingresses"],"verbs":["get"]}]}
  creationTimestamp: "2025-08-18T15:33:03Z"
  labels:
    app.kubernetes.io/name: keycloak-operator
  name: keycloak-operator-clusterrole
  resourceVersion: "22444256"
  uid: 642d0d98-5e09-405d-a897-89c2dc0cfdaa
rules:
- apiGroups:
  - apiextensions.k8s.io
  resources:
  - customresourcedefinitions
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - config.openshift.io
  resources:
  - ingresses
  verbs:
  - get
```

### Version

26.4.0

### Regression

- [ ] The issue is a regression

### Expected behavior

Permissions to work w/o patching

### Actual behavior

Fails with 

> 2025-09-30 17:50:07,029 WARN [io.jav.ope.pro.eve.EventProcessor] (ReconcilerExecutor-keycloakcontroller-54) Uncaught error during event processing ExecutionScope{ resource id: ResourceID{name='keycloak', namespace='auth'}, version: 22441437} - but ano ther reconciliation will be attempted because a superseding event has been received or another retry attempt is pending.: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1:443/apis/apiextensions.k8s.io /v1/customresourcedefinitions/servicemonitors.monitoring.coreos.com. Message: customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource " customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope. Received status: Status(apiVersion=v1, code=403, details=StatusDetails(causes=[], group=apiextensions.k8s.io, kind=customresourcedefinitions, name=servicemonitors.moni toring.coreos.com, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Forbidden, statu s=Failure, additionalProperties={}).

### How to Reproduce?

Installing per https://www.keycloak.org/operator/installation instructions, but into "auth" namespace instead of "keycloak" namespace. 

```bash
kubectl create namespace auth
kubectl -n auth apply -f https://raw.githubusercontent.com/keycloak/keycloak-k8s-resources/26.4.0/kubernetes/kubernetes.yml
kubectl patch clusterrolebinding keycloak-operator-clusterrole-binding --type='json' -p='[{"op": "replace", "path": "/subjects/0/namespace", "value":"auth"}]'
```


### Anything else?

_No response_

@shawkins 

Okay I did some more tests, and its somewhat weird edge case it seems. So I deleted the keycloak-operator deployment and installed it again via

```bash
kubectl apply -n auth -f https://raw.githubusercontent.com/keycloak/keycloak-k8s-resources/26.4.0/kubernetes/kubernetes.yml
```

and wait until the pod is up. I then get 

```
2025-10-01 14:54:02,812 INFO  [io.quarkus] (main) keycloak-operator 26.4.0 on JVM (powered by Quarkus 3.27.0) started in 14.611s. Listening on: http://0.0.0.0:8080
2025-10-01 14:54:02,892 INFO  [io.quarkus] (main) Profile prod activated. 
2025-10-01 14:54:02,892 INFO  [io.quarkus] (main) Installed features: [cdi, kubernetes, kubernetes-client, openshift-client, operator-sdk, smallrye-context-propagation, smallrye-health, vertx]
2025-10-01 14:54:04,499 WARN  [io.jav.ope.pro.eve.EventProcessor] (ReconcilerExecutor-keycloakcontroller-54) Uncaught error during event processing ExecutionScope{ resource id: ResourceID{name='keycloak', namespace='auth'}, version: 22932857} - but another reconciliation will be attempted because a superseding event has been received or another retry attempt is pending.: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1:443/apis/apiextensions.k8s.io/v1/customresourcedefinitions/servicemonitors.monitoring.coreos.com. Message: customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope. Received status: Status(apiVersion=v1, code=403, details=StatusDetails(causes=[], group=apiextensions.k8s.io, kind=customresourcedefinitions, name=servicemonitors.monitoring.coreos.com, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Forbidden, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:205)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleGet(OperationSupport.java:467)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handleGet(BaseOperation.java:792)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.requireFromServer(BaseOperation.java:193)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.get(BaseOperation.java:149)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.get(BaseOperation.java:98)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.CRDPresentActivationCondition$CRDPresentChecker.checkIfCRDPresent(CRDPresentActivationCondition.java:117)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.CRDPresentActivationCondition.isMet(CRDPresentActivationCondition.java:62)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.ConditionWithType.detailedIsMet(ConditionWithType.java:27)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.AbstractWorkflowExecutor.lambda$isConditionMet$2(AbstractWorkflowExecutor.java:144)
	at java.base/java.util.Optional.map(Optional.java:260)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.AbstractWorkflowExecutor.isConditionMet(AbstractWorkflowExecutor.java:142)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.WorkflowReconcileExecutor.handleReconcile(WorkflowReconcileExecutor.java:79)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.WorkflowReconcileExecutor.reconcile(WorkflowReconcileExecutor.java:28)
	at io.javaoperatorsdk.operator.processing.dependent.workflow.DefaultWorkflow.reconcile(DefaultWorkflow.java:100)
	at io.javaoperatorsdk.operator.processing.Controller.reconcileManagedWorkflow(Controller.java:469)
	at io.javaoperatorsdk.operator.api.reconciler.dependent.managed.DefaultManagedWorkflowAndDependentResourceContext.reconcileManagedWorkflow(DefaultManagedWorkflowAndDependentResourceContext.java:100)
	at org.keycloak.operator.controllers.KeycloakController.reconcile(KeycloakController.java:152)
	at org.keycloak.operator.controllers.KeycloakController.reconcile(KeycloakController.java:59)
	at org.keycloak.operator.controllers.KeycloakController_ClientProxy.reconcile(Unknown Source)
	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:161)
	at io.javaoperatorsdk.operator.processing.Controller$1.execute(Controller.java:117)
	at io.javaoperatorsdk.operator.api.monitoring.Metrics.timeControllerExecution(Metrics.java:159)
	at io.javaoperatorsdk.operator.processing.Controller.reconcile(Controller.java:116)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.reconcileExecution(ReconciliationDispatcher.java:153)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:130)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:97)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:68)
	at io.javaoperatorsdk.operator.processing.event.EventProcessor$ReconcilerExecutor.run(EventProcessor.java:475)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://10.96.0.1:443/apis/apiextensions.k8s.io/v1/customresourcedefinitions/servicemonitors.monitoring.coreos.com. Message: customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope. Received status: Status(apiVersion=v1, code=403, details=StatusDetails(causes=[], group=apiextensions.k8s.io, kind=customresourcedefinitions, name=servicemonitors.monitoring.coreos.com, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=customresourcedefinitions.apiextensions.k8s.io "servicemonitors.monitoring.coreos.com" is forbidden: User "system:serviceaccount:auth:keycloak-operator" cannot get resource "customresourcedefinitions" in API group "apiextensions.k8s.io" at the cluster scope, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Forbidden, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:642)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:622)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:582)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:141)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:51)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.vertx.VertxHttpRequest.lambda$consumeBytes$1(VertxHttpRequest.java:84)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:442)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:279)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:157)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:752)
	at io.vertx.core.impl.ContextImpl.lambda$execute$7(ContextImpl.java:329)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:472)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more
```
which is repeated 7 times before it stops retrying. 

Then I run

```bash
kubectl patch clusterrolebinding keycloak-operator-clusterrole-binding --type='json' -p='[{"op": "replace", "path": "/subjects/0/namespace", "value":"auth"}]'
```

and kill the keycloak-operator pod and get 

```

2025-10-01 14:57:15,040 INFO  [io.quarkus] (main) keycloak-operator 26.4.0 on JVM (powered by Quarkus 3.27.0) started in 15.344s. Listening on: http://0.0.0.0:8080
2025-10-01 14:57:15,040 INFO  [io.quarkus] (main) Profile prod activated. 
2025-10-01 14:57:15,041 INFO  [io.quarkus] (main) Installed features: [cdi, kubernetes, kubernetes-client, openshift-client, operator-sdk, smallrye-context-propagation, smallrye-health, vertx]
2025-10-01 14:57:19,052 ERROR [io.jav.ope.pro.eve.ReconciliationDispatcher] (ReconcilerExecutor-keycloakcontroller-62) updateErrorStatus failed for resource: a86407e2-7e87-43a7-9465-52104c92c8ba with version: 22934977 for error Failure executing: PATCH at: https://10.96.0.1:443/apis/k8s.keycloak.org/v2alpha1/namespaces/auth/keycloaks/keycloak/status?fieldManager=keycloakcontroller&force=true. Message: Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=k8s.keycloak.org, kind=keycloaks, name=keycloak, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}).: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://10.96.0.1:443/apis/k8s.keycloak.org/v2alpha1/namespaces/auth/keycloaks/keycloak/status?fieldManager=keycloakcontroller&force=true. Message: Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=k8s.keycloak.org, kind=keycloaks, name=keycloak, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.KubernetesClientException.copyAsCause(KubernetesClientException.java:205)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.waitForResult(OperationSupport.java:507)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handleResponse(OperationSupport.java:524)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:419)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.handlePatch(OperationSupport.java:397)
	at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.handlePatch(BaseOperation.java:764)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.lambda$patch$2(HasMetadataOperation.java:231)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:236)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:251)
	at io.fabric8.kubernetes.client.dsl.internal.HasMetadataOperation.patch(HasMetadataOperation.java:44)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher$CustomResourceFacade.patchStatus(ReconciliationDispatcher.java:462)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleErrorStatusHandler(ReconciliationDispatcher.java:215)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleReconcile(ReconciliationDispatcher.java:132)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleDispatch(ReconciliationDispatcher.java:97)
	at io.javaoperatorsdk.operator.processing.event.ReconciliationDispatcher.handleExecution(ReconciliationDispatcher.java:68)
	at io.javaoperatorsdk.operator.processing.event.EventProcessor$ReconcilerExecutor.run(EventProcessor.java:475)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: PATCH at: https://10.96.0.1:443/apis/k8s.keycloak.org/v2alpha1/namespaces/auth/keycloaks/keycloak/status?fieldManager=keycloakcontroller&force=true. Message: Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=k8s.keycloak.org, kind=keycloaks, name=keycloak, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}).
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:642)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.requestFailure(OperationSupport.java:622)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.assertResponseCode(OperationSupport.java:582)
	at io.fabric8.kubernetes.client.dsl.internal.OperationSupport.lambda$handleResponse$0(OperationSupport.java:549)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:646)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.StandardHttpClient.lambda$completeOrCancel$10(StandardHttpClient.java:141)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.http.ByteArrayBodyHandler.onBodyDone(ByteArrayBodyHandler.java:51)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at io.fabric8.kubernetes.client.vertx.VertxHttpRequest.lambda$consumeBytes$1(VertxHttpRequest.java:84)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:270)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:252)
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:76)
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:442)
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:279)
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:157)
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:752)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:327)
	at io.vertx.core.impl.ContextImpl.execute(ContextImpl.java:307)
	at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:983)
	at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:855)
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:817)
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:159)
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1519)
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1377)
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1428)
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:530)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:469)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:796)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:732)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:658)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:998)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	... 1 more

2025-10-01 14:57:19,058 INFO  [io.jav.ope.pro.eve.EventProcessor] (ReconcilerExecutor-keycloakcontroller-62) Resource Kubernetes Resource Creator/Update Conflict during reconciliation. Message: Failure executing: PATCH at: https://10.96.0.1:443/apis/k8s.keycloak.org/v2alpha1/namespaces/auth/keycloaks/keycloak/status?fieldManager=keycloakcontroller&force=true. Message: Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again. Received status: Status(apiVersion=v1, code=409, details=StatusDetails(causes=[], group=k8s.keycloak.org, kind=keycloaks, name=keycloak, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=Operation cannot be fulfilled on keycloaks.k8s.keycloak.org "keycloak": the object has been modified; please apply your changes to the latest version and try again, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=Conflict, status=Failure, additionalProperties={}). Resource name: null
```

then i run 

```bash
kubectl edit clusterrole keycloak-operator-clusterrole
```

add "list/watch" and kill the pod again

and it logs

```
2025-10-01 15:14:35,388 INFO  [io.quarkus] (main) keycloak-operator 26.4.0 on JVM (powered by Quarkus 3.27.0) started in 14.701s. Listening on: http://0.0.0.0:8080
2025-10-01 15:14:35,388 INFO  [io.quarkus] (main) Profile prod activated.
2025-10-01 15:14:35,389 INFO  [io.quarkus] (main) Installed features: [cdi, kubernetes, kubernetes-client, openshift-client, operator-sdk, smallrye-context-propagation, smallrye-health, vertx]
```

without any exceptions.

However, removing the two verbs again and killing the pod still runs w/o exceptions.

So I removed the operator deployment again and ran

```
kubectl apply -n auth -f https://raw.githubusercontent.com/keycloak/keycloak-k8s-resources/26.4.0/kubernetes/kubernetes.yml && \
    kubectl patch clusterrolebinding keycloak-operator-clusterrole-binding --type='json' -p='[{"op": "replace", "path": "/subjects/0/namespace", "value":"auth"}]'
```

instead and it seem to run fine too. Odd.

--- a/.github/workflows/operator-ci.yml
+++ b/.github/workflows/operator-ci.yml
@@ -206,6 +206,7 @@ jobs:
         run: |
           kubectl apply -f src/test/resources/service-monitor-crds.yml
           ./scripts/check-crd-installed.sh servicemonitors
+          kubectl delete pod -l name=keycloak-operator
 
       - name: Deploy an example Keycloak with ServiceMonitor
         working-directory: operator/scripts

--- a/docs/guides/operator/installation.adoc
+++ b/docs/guides/operator/installation.adoc
@@ -125,6 +125,8 @@ execute the following commands:
 kubectl create namespace custom-namespace
 kubectl -n custom-namespace apply -f https://raw.githubusercontent.com/keycloak/keycloak-k8s-resources/{version}/kubernetes/kubernetes.yml
 kubectl patch clusterrolebinding keycloak-operator-clusterrole-binding --type='json' -p='[{"op": "replace", "path": "/subjects/0/namespace", "value":"custom-namespace"}]'
+# if you have existing keycloaks, restart the operator after patching the clusterrolebinding
+kubectl rollout restart -n custom-namespace Deployment/keycloak-operator
 ----
 </@profile.ifCommunity>
 

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
@@ -19,6 +19,7 @@ package org.keycloak.operator.controllers;
 import io.fabric8.kubernetes.api.model.ContainerState;
 import io.fabric8.kubernetes.api.model.ContainerStateWaiting;
 import io.fabric8.kubernetes.api.model.ContainerStatus;
+import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.PodSpec;
 import io.fabric8.kubernetes.api.model.PodStatus;
 import io.fabric8.kubernetes.api.model.apps.StatefulSet;
@@ -34,6 +35,7 @@ import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
 import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
 import io.javaoperatorsdk.operator.api.reconciler.Workflow;
 import io.javaoperatorsdk.operator.api.reconciler.dependent.Dependent;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
 import io.javaoperatorsdk.operator.processing.dependent.workflow.CRDPresentActivationCondition;
 import io.javaoperatorsdk.operator.processing.event.source.EventSource;
 import io.quarkus.logging.Log;
@@ -51,6 +53,7 @@ import org.keycloak.operator.crds.v2alpha1.deployment.spec.HostnameSpec;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.HostnameSpecBuilder;
 import org.keycloak.operator.update.UpdateLogicFactory;
 
+import java.net.HttpURLConnection;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Optional;
@@ -67,11 +70,34 @@ import java.util.concurrent.TimeUnit;
         @Dependent(type = KeycloakNetworkPolicyDependentResource.class, reconcilePrecondition = KeycloakNetworkPolicyDependentResource.EnabledCondition.class),
         @Dependent(
               type = KeycloakServiceMonitorDependentResource.class,
-              activationCondition = CRDPresentActivationCondition.class,
+              activationCondition = KeycloakController.ExceptionAwareCRDPresentActivationCondition.class,
               reconcilePrecondition = KeycloakServiceMonitorDependentResource.ReconcilePrecondition.class
         ),
     })
 public class KeycloakController implements Reconciler<Keycloak> {
+
+    public static class ExceptionAwareCRDPresentActivationCondition<R extends HasMetadata, P extends HasMetadata> extends CRDPresentActivationCondition<R, P> {
+
+        boolean logged = false;
+
+        @Override
+        public boolean isMet(DependentResource<R, P> dependentResource, P primary, Context<P> context) {
+            try {
+                return super.isMet(dependentResource, primary, context);
+            } catch (KubernetesClientException e) {
+                if (e.getCode() == HttpURLConnection.HTTP_FORBIDDEN) {
+                    if (!logged) {
+                        logged = true;
+                        Log.info(
+                                "Operator service account lacks permission for %s. If that is added later, you will need to restart the operator."
+                                        .formatted(HasMetadata.getFullResourceName(dependentResource.resourceType())));
+                    }
+                    return false;
+                }
+                throw e; // allow the default handling, but realistically it's better to return false, than allow retries to be exhausted
+            }
+        }
+    }
 
     public static final String OPENSHIFT_DEFAULT = "openshift-default";
 

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakController.java
@@ -19,7 +19,6 @@ package org.keycloak.operator.controllers;
 import io.fabric8.kubernetes.api.model.ContainerState;
 import io.fabric8.kubernetes.api.model.ContainerStateWaiting;
 import io.fabric8.kubernetes.api.model.ContainerStatus;
-import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.PodSpec;
 import io.fabric8.kubernetes.api.model.PodStatus;
 import io.fabric8.kubernetes.api.model.apps.StatefulSet;
@@ -35,8 +34,6 @@ import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
 import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
 import io.javaoperatorsdk.operator.api.reconciler.Workflow;
 import io.javaoperatorsdk.operator.api.reconciler.dependent.Dependent;
-import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
-import io.javaoperatorsdk.operator.processing.dependent.workflow.CRDPresentActivationCondition;
 import io.javaoperatorsdk.operator.processing.event.source.EventSource;
 import io.quarkus.logging.Log;
 import jakarta.inject.Inject;
@@ -53,7 +50,6 @@ import org.keycloak.operator.crds.v2alpha1.deployment.spec.HostnameSpec;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.HostnameSpecBuilder;
 import org.keycloak.operator.update.UpdateLogicFactory;
 
-import java.net.HttpURLConnection;
 import java.util.Comparator;
 import java.util.List;
 import java.util.Optional;
@@ -70,34 +66,10 @@ import java.util.concurrent.TimeUnit;
         @Dependent(type = KeycloakNetworkPolicyDependentResource.class, reconcilePrecondition = KeycloakNetworkPolicyDependentResource.EnabledCondition.class),
         @Dependent(
               type = KeycloakServiceMonitorDependentResource.class,
-              activationCondition = KeycloakController.ExceptionAwareCRDPresentActivationCondition.class,
-              reconcilePrecondition = KeycloakServiceMonitorDependentResource.ReconcilePrecondition.class
+              activationCondition = KeycloakServiceMonitorDependentResource.ActivationCondition.class
         ),
     })
 public class KeycloakController implements Reconciler<Keycloak> {
-
-    public static class ExceptionAwareCRDPresentActivationCondition<R extends HasMetadata, P extends HasMetadata> extends CRDPresentActivationCondition<R, P> {
-
-        boolean logged = false;
-
-        @Override
-        public boolean isMet(DependentResource<R, P> dependentResource, P primary, Context<P> context) {
-            try {
-                return super.isMet(dependentResource, primary, context);
-            } catch (KubernetesClientException e) {
-                if (e.getCode() == HttpURLConnection.HTTP_FORBIDDEN) {
-                    if (!logged) {
-                        logged = true;
-                        Log.info(
-                                "Operator service account lacks permission for %s. If that is added later, you will need to restart the operator."
-                                        .formatted(HasMetadata.getFullResourceName(dependentResource.resourceType())));
-                    }
-                    return false;
-                }
-                throw e; // allow the default handling, but realistically it's better to return false, than allow retries to be exhausted
-            }
-        }
-    }
 
     public static final String OPENSHIFT_DEFAULT = "openshift-default";
 
@@ -256,6 +228,10 @@ public class KeycloakController implements Reconciler<Keycloak> {
         }
 
         distConfigurator.validateOptions(keycloakCR, status);
+
+        context.managedWorkflowAndDependentResourceContext()
+                .get(KeycloakServiceMonitorDependentResource.SERVICE_MONITOR_WARNING, String.class)
+                .ifPresent(status::addWarningMessage);
     }
 
     public static boolean isRolling(StatefulSet existingDeployment) {

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakDeploymentDependentResource.java
@@ -75,6 +75,7 @@ import java.util.stream.Stream;
 
 import static org.keycloak.operator.Utils.addResources;
 import static org.keycloak.operator.controllers.KeycloakDistConfigurator.getKeycloakOptionEnvVarName;
+import static org.keycloak.operator.crds.v2alpha1.CRDUtils.LEGACY_MANAGEMENT_ENABLED;
 import static org.keycloak.operator.crds.v2alpha1.CRDUtils.isTlsConfigured;
 import static org.keycloak.operator.crds.v2alpha1.deployment.spec.TracingSpec.convertTracingAttributesToString;
 
@@ -83,7 +84,6 @@ import static org.keycloak.operator.crds.v2alpha1.deployment.spec.TracingSpec.co
 )
 public class KeycloakDeploymentDependentResource extends CRUDKubernetesDependentResource<StatefulSet, Keycloak> {
 
-    public static final String HTTP_MANAGEMENT_HEALTH_ENABLED = "http-management-health-enabled";
     public static final String HTTP_MANAGEMENT_SCHEME = "http-management-scheme";
 
     public static final String POD_IP = "POD_IP";
@@ -356,8 +356,7 @@ public class KeycloakDeploymentDependentResource extends CRUDKubernetesDependent
         // Set bind address as this is required for JGroups to form a cluster in IPv6 environments
         containerBuilder.addToArgs(0, "-Djgroups.bind.address=$(%s)".formatted(POD_IP));
 
-        var healthEnabled = readConfigurationValue(HTTP_MANAGEMENT_HEALTH_ENABLED, keycloakCR, context).map(Boolean::valueOf).orElse(true);
-        ManagementEndpoint endpoint = managementEndpoint(keycloakCR, context, healthEnabled);
+        ManagementEndpoint endpoint = managementEndpoint(keycloakCR, context, true);
 
         // probes
         var readinessOptionalSpec = Optional.ofNullable(keycloakCR.getSpec().getReadinessProbeSpec());
@@ -592,7 +591,7 @@ public class KeycloakDeploymentDependentResource extends CRUDKubernetesDependent
         return keycloak.getMetadata().getName();
     }
 
-    private static Optional<String> readConfigurationValue(String key, Keycloak keycloakCR, Context<Keycloak> context) {
+    static Optional<String> readConfigurationValue(String key, Keycloak keycloakCR, Context<Keycloak> context) {
         return Optional.ofNullable(keycloakCR.getSpec()).map(KeycloakSpec::getAdditionalOptions)
                 .flatMap(l -> l.stream().filter(sc -> sc.getName().equals(key)).findFirst().map(serverConfigValue -> {
             if (serverConfigValue.getValue() != null) {
@@ -644,20 +643,27 @@ public class KeycloakDeploymentDependentResource extends CRUDKubernetesDependent
         toUpdate.getMetadata().getAnnotations().put(Constants.KEYCLOAK_UPDATE_REVISION_ANNOTATION, revision);
     }
 
-    record ManagementEndpoint(String relativePath, String protocol, int port) {}
+    record ManagementEndpoint(String relativePath, String protocol, int port, String portName) {}
 
-    static ManagementEndpoint managementEndpoint(Keycloak keycloakCR, Context<Keycloak> context, boolean useMgmtProtocolPort) {
+    static ManagementEndpoint managementEndpoint(Keycloak keycloakCR, Context<Keycloak> context, boolean health) {
         boolean tls = isTlsConfigured(keycloakCR);
         String protocol = tls ? "HTTPS" : "HTTP";
         int port;
+        String portName;
 
-        if (useMgmtProtocolPort) {
+        var legacy = readConfigurationValue(LEGACY_MANAGEMENT_ENABLED, keycloakCR, context).map(Boolean::valueOf).orElse(false);
+
+        var healthManagementEnabled = readConfigurationValue(CRDUtils.HTTP_MANAGEMENT_HEALTH_ENABLED, keycloakCR, context).map(Boolean::valueOf).orElse(true);
+
+        if (!legacy && (!health || healthManagementEnabled)) {
             port = HttpManagementSpec.managementPort(keycloakCR);
+            portName = Constants.KEYCLOAK_MANAGEMENT_PORT_NAME;
             if (readConfigurationValue(HTTP_MANAGEMENT_SCHEME, keycloakCR, context).filter("http"::equals).isPresent()) {
                 protocol = "HTTP";
             }
         } else {
             port = tls ? HttpSpec.httpsPort(keycloakCR) : HttpSpec.httpPort(keycloakCR);
+            portName = tls ? Constants.KEYCLOAK_HTTPS_PORT_NAME : Constants.KEYCLOAK_HTTP_PORT_NAME;
         }
 
         var relativePath = readConfigurationValue(Constants.KEYCLOAK_HTTP_MANAGEMENT_RELATIVE_PATH_KEY, keycloakCR, context)
@@ -665,6 +671,6 @@ public class KeycloakDeploymentDependentResource extends CRUDKubernetesDependent
               .map(path -> !path.endsWith("/") ? path + "/" : path)
               .orElse("/");
 
-        return new ManagementEndpoint(relativePath, protocol, port);
+        return new ManagementEndpoint(relativePath, protocol, port, portName);
     }
 }

--- a/operator/src/main/java/org/keycloak/operator/controllers/KeycloakServiceMonitorDependentResource.java
+++ b/operator/src/main/java/org/keycloak/operator/controllers/KeycloakServiceMonitorDependentResource.java
@@ -1,16 +1,19 @@
 package org.keycloak.operator.controllers;
 
 import static org.keycloak.operator.controllers.KeycloakDeploymentDependentResource.managementEndpoint;
-import static org.keycloak.operator.crds.v2alpha1.CRDUtils.LEGACY_MANAGEMENT_ENABLED;
 import static org.keycloak.operator.crds.v2alpha1.CRDUtils.METRICS_ENABLED;
 import static org.keycloak.operator.crds.v2alpha1.CRDUtils.configuredOptions;
+
+import java.net.HttpURLConnection;
 
 import org.keycloak.operator.Constants;
 import org.keycloak.operator.Utils;
 import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.ServiceMonitorSpec;
 
-import io.fabric8.kubernetes.api.model.networking.v1.Ingress;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.WatcherException;
 import io.fabric8.openshift.api.model.monitoring.v1.ServiceMonitor;
 import io.fabric8.openshift.api.model.monitoring.v1.ServiceMonitorBuilder;
 import io.javaoperatorsdk.operator.api.config.informer.Informer;
@@ -19,28 +22,75 @@ import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
 import io.javaoperatorsdk.operator.processing.dependent.kubernetes.CRUDKubernetesDependentResource;
 import io.javaoperatorsdk.operator.processing.dependent.kubernetes.KubernetesDependent;
 import io.javaoperatorsdk.operator.processing.dependent.workflow.Condition;
+import io.quarkus.logging.Log;
 
 @KubernetesDependent(
       informer = @Informer(labelSelector = Constants.DEFAULT_LABELS_AS_STRING)
 )
 public class KeycloakServiceMonitorDependentResource extends CRUDKubernetesDependentResource<ServiceMonitor, Keycloak> {
 
-    public static class ReconcilePrecondition implements Condition<Ingress, Keycloak> {
+    public static final String WARN_METRICS_NOT_ENABLED = "A ServiceMonitor will not be created because `metrics-enabled` is not true.";
+    public static final String WARN_CRD_NOT_INSTALLED = "A ServiceMonitor will not be created because the ServiceMonitor CRD is not installed.";
+
+    static String SERVICE_MONITOR_WARNING = "ServiceMonitorWarning";
+
+    volatile Boolean crdInstalled;
+
+    public static class ActivationCondition implements Condition<ServiceMonitor, Keycloak> {
+
         @Override
-        public boolean isMet(DependentResource<Ingress, Keycloak> dependentResource, Keycloak primary,
-                             Context<Keycloak> context) {
-            var opts = configuredOptions(primary);
-            if (Boolean.parseBoolean(opts.get(LEGACY_MANAGEMENT_ENABLED)) || !Boolean.parseBoolean(opts.getOrDefault(METRICS_ENABLED, "false"))) {
+        public boolean isMet(DependentResource<ServiceMonitor, Keycloak> dependentResource, Keycloak primary, Context<Keycloak> context) {
+            if (!ServiceMonitorSpec.get(primary).isEnabled()) {
                 return false;
             }
 
-            return ServiceMonitorSpec.get(primary).isEnabled();
+            var opts = configuredOptions(primary);
+
+            if (!Boolean.parseBoolean(opts.getOrDefault(METRICS_ENABLED, "false"))) {
+                context.managedWorkflowAndDependentResourceContext().put(SERVICE_MONITOR_WARNING, WARN_METRICS_NOT_ENABLED);
+                return false;
+            }
+
+            if (!isCRDInstalled(dependentResource, context, (KeycloakServiceMonitorDependentResource)dependentResource)) {
+                context.managedWorkflowAndDependentResourceContext().put(SERVICE_MONITOR_WARNING, WARN_CRD_NOT_INSTALLED);
+                return false;
+            }
+
+            return true;
+        }
+
+        private boolean isCRDInstalled(DependentResource<ServiceMonitor, Keycloak> dependentResource,
+                Context<Keycloak> context, KeycloakServiceMonitorDependentResource serviceMonitorDependentResource) {
+            if (serviceMonitorDependentResource.crdInstalled != null) {
+                return serviceMonitorDependentResource.crdInstalled;
+            }
+            Watcher<ServiceMonitor> dummyWatcher = new Watcher<ServiceMonitor>() {
+
+                @Override
+                public void eventReceived(Action action, ServiceMonitor resource) {
+                }
+
+                @Override
+                public void onClose(WatcherException cause) {
+                }
+            };
+            try (var watch = context.getClient().resources(dependentResource.resourceType()).watch(dummyWatcher)) {
+                serviceMonitorDependentResource.crdInstalled = true;
+                return true;
+            } catch (KubernetesClientException e) {
+                if (e.getCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+                    Log.warn("ServiceMonitors will not be managed by the operator as the CRD for ServiceMonitors not installed. If the CRD is installed later, the operator will need to be restarted.");
+                    serviceMonitorDependentResource.crdInstalled = false;
+                    return false;
+                }
+                throw e;
+            }
         }
     }
 
     @Override
     protected ServiceMonitor desired(Keycloak primary, Context<Keycloak> context) {
-        var endpoint = managementEndpoint(primary, context, true);
+        var endpoint = managementEndpoint(primary, context, false);
         var meta = primary.getMetadata();
         var spec = ServiceMonitorSpec.get(primary);
         return new ServiceMonitorBuilder()
@@ -60,7 +110,7 @@ public class KeycloakServiceMonitorDependentResource extends CRUDKubernetesDepen
                 .addNewEndpoint()
                   .withInterval(spec.getInterval())
                   .withPath(endpoint.relativePath() + "metrics")
-                  .withPort(Constants.KEYCLOAK_MANAGEMENT_PORT_NAME)
+                  .withPort(endpoint.portName())
                   .withScheme(endpoint.protocol().toLowerCase())
                   .withScrapeTimeout(spec.getScrapeTimeout())
                   .withNewTlsConfig()

--- a/operator/src/main/java/org/keycloak/operator/crds/v2alpha1/CRDUtils.java
+++ b/operator/src/main/java/org/keycloak/operator/crds/v2alpha1/CRDUtils.java
@@ -21,10 +21,8 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.Objects;
 import java.util.Optional;
 import java.util.function.Predicate;
-import java.util.stream.Stream;
 
 import org.keycloak.operator.Constants;
 import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
@@ -47,6 +45,7 @@ import io.javaoperatorsdk.operator.api.reconciler.Context;
  */
 public final class CRDUtils {
     private static final String HEALTH_ENABLED = "health-enabled";
+    public static final String HTTP_MANAGEMENT_HEALTH_ENABLED = "http-management-health-enabled";
     public static final String METRICS_ENABLED = "metrics-enabled";
     public static final String LEGACY_MANAGEMENT_ENABLED = "legacy-observability-interface";
 
@@ -71,11 +70,8 @@ public final class CRDUtils {
             return false;
         }
 
-        // Only metrics and health use the management endpoint.
-        return Stream.of(METRICS_ENABLED, HEALTH_ENABLED)
-                .map(options::get)
-                .filter(Objects::nonNull)
-                .anyMatch(Boolean::parseBoolean);
+        return Boolean.parseBoolean(options.get(METRICS_ENABLED)) || (Boolean.parseBoolean(options.get(HEALTH_ENABLED))
+                && Boolean.parseBoolean(options.getOrDefault(HTTP_MANAGEMENT_HEALTH_ENABLED, Boolean.toString(true))));
     }
 
     public static Map<String, String> configuredOptions(Keycloak keycloak) {

--- a/operator/src/main/kubernetes/kubernetes.yml
+++ b/operator/src/main/kubernetes/kubernetes.yml
@@ -94,12 +94,6 @@ metadata:
     app.kubernetes.io/name: keycloak-operator
 rules:
   - apiGroups:
-      - apiextensions.k8s.io
-    resources:
-      - customresourcedefinitions
-    verbs:
-      - get
-  - apiGroups:
       - config.openshift.io
     resources:
       - ingresses

--- a/operator/src/test/java/org/keycloak/operator/testsuite/integration/ServiceMonitorTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/integration/ServiceMonitorTest.java
@@ -8,9 +8,12 @@ import org.awaitility.Awaitility;
 import org.junit.jupiter.api.Assumptions;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
+import org.keycloak.operator.controllers.KeycloakServiceMonitorDependentResource;
 import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusCondition;
 import org.keycloak.operator.crds.v2alpha1.deployment.ValueOrSecret;
 import org.keycloak.operator.crds.v2alpha1.deployment.spec.ServiceMonitorSpecBuilder;
+import org.keycloak.operator.testsuite.utils.CRAssert;
 import org.keycloak.operator.testsuite.utils.K8sUtils;
 
 import io.fabric8.kubernetes.client.KubernetesClient;
@@ -24,18 +27,23 @@ public class ServiceMonitorTest extends BaseOperatorTest {
     @Test
     public void testServiceMonitorDisabledNoMetrics() {
         Assumptions.assumeTrue(isServiceMonitorAvailable(k8sclient));
-        var kc = getTestKeycloakDeployment(true, false);;
+        var kc = getTestKeycloakDeployment(true, false);
         kc.getSpec().setAdditionalOptions(List.of(new ValueOrSecret("metrics-enabled", "false")));
         K8sUtils.deployKeycloak(k8sclient, kc, true);
 
         ServiceMonitor sm = getServiceMonitor(kc);
         assertThat(sm).isNull();
+
+        CRAssert.assertKeycloakStatusCondition(
+                k8sclient.resources(Keycloak.class).withName(kc.getMetadata().getName()).get(),
+                KeycloakStatusCondition.HAS_ERRORS, false,
+                KeycloakServiceMonitorDependentResource.WARN_METRICS_NOT_ENABLED);
     }
 
     @Test
     public void testServiceMonitorCreatedWithMetricsEnabled() {
         Assumptions.assumeTrue(isServiceMonitorAvailable(k8sclient));
-        var kc = getTestKeycloakDeployment(true, false);;
+        var kc = getTestKeycloakDeployment(true, false);
         K8sUtils.deployKeycloak(k8sclient, kc, true);
 
         Awaitility.await().untilAsserted(() -> {
@@ -48,7 +56,7 @@ public class ServiceMonitorTest extends BaseOperatorTest {
     @Test
     public void testServiceMonitorDisabledExplicitly() {
         Assumptions.assumeTrue(isServiceMonitorAvailable(k8sclient));
-        var kc = getTestKeycloakDeployment(true, false);;
+        var kc = getTestKeycloakDeployment(true, false);
         kc.getSpec().setServiceMonitorSpec(
               new ServiceMonitorSpecBuilder()
                     .withEnabled(false)
@@ -61,20 +69,24 @@ public class ServiceMonitorTest extends BaseOperatorTest {
     }
 
     @Test
-    public void testServiceMonitorDisabledLegacyManagement() {
+    public void testServiceMonitorLegacyManagement() {
         Assumptions.assumeTrue(isServiceMonitorAvailable(k8sclient));
-        var kc = getTestKeycloakDeployment(true, false);;
-        kc.getSpec().setAdditionalOptions(List.of(new ValueOrSecret("legacy-observability-interface", "true")));
+        var kc = getTestKeycloakDeployment(true, false);
+        kc.getSpec().getAdditionalOptions().add(new ValueOrSecret("legacy-observability-interface", "true"));
         K8sUtils.deployKeycloak(k8sclient, kc, true);
 
-        ServiceMonitor sm = getServiceMonitor(kc);
-        assertThat(sm).isNull();
+        Awaitility.await().untilAsserted(() -> {
+            var sm = getServiceMonitor(kc);
+            assertThat(sm).isNotNull();
+            assertThat(sm.getSpec().getEndpoints()).hasSize(1);
+            assertThat(sm.getSpec().getEndpoints().get(0).getPort()).isEqualTo("https");
+        });
     }
 
     @Test
     public void testServiceMonitorConfigProperties() {
         Assumptions.assumeTrue(isServiceMonitorAvailable(k8sclient));
-        var kc = getTestKeycloakDeployment(true, false);;
+        var kc = getTestKeycloakDeployment(true, false);
         kc.getSpec().setServiceMonitorSpec(
               new ServiceMonitorSpecBuilder()
                     .withInterval("1s")
@@ -92,14 +104,14 @@ public class ServiceMonitorTest extends BaseOperatorTest {
         });
     }
 
-    private ServiceMonitor getServiceMonitor(Keycloak kc) {
+    static ServiceMonitor getServiceMonitor(Keycloak kc) {
         return k8sclient.resources(ServiceMonitor.class)
               .inNamespace(kc.getMetadata().getNamespace())
               .withName(kc.getMetadata().getName())
               .get();
     }
 
-    private boolean isServiceMonitorAvailable(KubernetesClient client) {
+    static boolean isServiceMonitorAvailable(KubernetesClient client) {
         return client
               .apiextensions()
               .v1()

--- a/operator/src/test/java/org/keycloak/operator/testsuite/integration/ServiceMonitorUninstalledTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/integration/ServiceMonitorUninstalledTest.java
@@ -1,0 +1,43 @@
+package org.keycloak.operator.testsuite.integration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.concurrent.TimeUnit;
+
+import org.junit.jupiter.api.Assumptions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.keycloak.operator.controllers.KeycloakServiceMonitorDependentResource;
+import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
+import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakStatusCondition;
+import org.keycloak.operator.testsuite.utils.CRAssert;
+import org.keycloak.operator.testsuite.utils.K8sUtils;
+
+import io.fabric8.openshift.api.model.monitoring.v1.ServiceMonitor;
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class ServiceMonitorUninstalledTest extends BaseOperatorTest {
+
+    @BeforeAll
+    public static void beforeAll() {
+        k8sclient.apiextensions().v1().customResourceDefinitions().withName(new ServiceMonitor().getFullResourceName())
+                .withTimeout(10, TimeUnit.SECONDS).delete();
+    }
+
+    @Test
+    public void testServiceMonitorNoCRD() {
+        Assumptions.assumeFalse(ServiceMonitorTest.isServiceMonitorAvailable(k8sclient));
+        var kc = getTestKeycloakDeployment(true, false);
+        K8sUtils.deployKeycloak(k8sclient, kc, true);
+
+        ServiceMonitor sm = ServiceMonitorTest.getServiceMonitor(kc);
+        assertThat(sm).isNull();
+
+        CRAssert.assertKeycloakStatusCondition(
+                k8sclient.resources(Keycloak.class).withName(kc.getMetadata().getName()).get(),
+                KeycloakStatusCondition.HAS_ERRORS, false,
+                KeycloakServiceMonitorDependentResource.WARN_CRD_NOT_INSTALLED);
+    }
+
+}

--- a/operator/src/test/java/org/keycloak/operator/testsuite/unit/PodTemplateTest.java
+++ b/operator/src/test/java/org/keycloak/operator/testsuite/unit/PodTemplateTest.java
@@ -57,6 +57,7 @@ import org.keycloak.operator.controllers.KeycloakDistConfigurator;
 import org.keycloak.operator.controllers.KeycloakRealmImportJobDependentResource;
 import org.keycloak.operator.controllers.KeycloakUpdateJobDependentResource;
 import org.keycloak.operator.controllers.WatchedResources;
+import org.keycloak.operator.crds.v2alpha1.CRDUtils;
 import org.keycloak.operator.crds.v2alpha1.deployment.Keycloak;
 import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakBuilder;
 import org.keycloak.operator.crds.v2alpha1.deployment.KeycloakSpecBuilder;
@@ -416,7 +417,7 @@ public class PodTemplateTest {
     @Test
     public void testHealthOnMain() {
         var result = getDeployment(null, new StatefulSet(),
-                spec -> spec.withAdditionalOptions(new ValueOrSecret(KeycloakDeploymentDependentResource.HTTP_MANAGEMENT_HEALTH_ENABLED, "false")))
+                spec -> spec.withAdditionalOptions(new ValueOrSecret(CRDUtils.HTTP_MANAGEMENT_HEALTH_ENABLED, "false")))
                 .getSpec()
                 .getTemplate()
                 .getSpec()

